File: tools/math_dsl/persistent_multiplier.py
"""W244 wave 19 — closed-form analytical model for `persistent_multiplier`.

Industry pattern (Quickspin Sticky Bandits, Pragmatic Mighty Wild,
NetEnt Money Vault FS, ELK Bompergo escalator, BTG Extra Chilli mega
multiplier):

  Multiplier dynamics during FS
  ----------------------------
    Multiplier starts at `initial_multiplier` (typical 1×) and persists
    across FS spins. On each FS spin, with probability `p_bump_per_spin`
    the multiplier increases by `bump_increment` (typical +1). Multiplier
    optionally capped at `max_multiplier`.

    Final per-FS-spin base award is multiplied by the CURRENT multiplier
    at spin time.

  Closed-form RTP contribution
  ----------------------------
    Per-FS-spin pay (without multiplier): `base_pay_per_spin_x_bet`.

    Multiplier path across FS:
      m_t = initial + (bump_increment × #_bumps_so_far)
      bumps_so_far ∼ Binomial(t-1, p_bump_per_spin) on spin t.

    E[multiplier on spin t] = initial + bump_increment × (t-1) × p_bump
      capped at max_multiplier.

    Total FS award:
      E[fs_total] = base_pay × sum_t E[multiplier_t]

    For uncapped:
      sum_t E[multiplier_t] = T × initial +
                              bump × p_bump × T × (T-1) / 2

    For capped: piecewise — analytical up to cap-hit spin, then constant.
    We use exact DP over multiplier state for tractable cap.

  Per-base-spin RTP
  -----------------
    RTP[persistent_multiplier_FS] = fs_trigger_p × E[fs_total]

Pure-stdlib. Used by:
  * `tools.math_dsl.compile` for IR emission
  * `tools/build_persistent_multiplier_kernel.py` for acceptance artefact
  * `tools/tests/test_w244_persistent_multiplier_kernel.py` for closed-form pin
"""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class PersistentMultiplierParams:
    """Closed-form model inputs."""
    fs_trigger_p: float                      # P(FS triggers per base spin)
    fs_initial_spins: int                    # typical 10
    base_pay_per_spin_x_bet: float           # avg per-FS-spin pay (no mult)
    initial_multiplier: float = 1.0          # starting multiplier
    bump_increment: float = 1.0              # per-bump increment
    p_bump_per_spin: float = 0.30            # P(bump) per FS spin
    max_multiplier: float | None = None      # None = uncapped

    def __post_init__(self):
        if not (0.0 <= self.fs_trigger_p <= 1.0):
            raise ValueError(f"fs_trigger_p {self.fs_trigger_p} outside [0,1]")
        if self.fs_initial_spins < 1:
            raise ValueError("fs_initial_spins must be ≥ 1")
        if self.base_pay_per_spin_x_bet < 0:
            raise ValueError("base_pay_per_spin_x_bet must be ≥ 0")
        if self.initial_multiplier < 0:
            raise ValueError("initial_multiplier must be ≥ 0")
        if self.bump_increment < 0:
            raise ValueError("bump_increment must be ≥ 0")
        if not (0.0 <= self.p_bump_per_spin <= 1.0):
            raise ValueError(
                f"p_bump_per_spin {self.p_bump_per_spin} outside [0,1]"
            )
        if self.max_multiplier is not None and (
            self.max_multiplier < self.initial_multiplier
        ):
            raise ValueError(
                f"max_multiplier {self.max_multiplier} must exceed "
                f"initial_multiplier {self.initial_multiplier}"
            )


def expected_multiplier_at_spin(params: PersistentMultiplierParams, t: int) -> float:
    """E[multiplier on FS spin t (1-indexed)], with cap.

    Without cap:
      E[m_t] = initial + bump × p_bump × (t - 1)

    With cap: clamp expectation at max_multiplier (approximation — exact
    handling requires per-bump-count DP since cap-hitting flips the bump
    distribution. We use the exact DP via `_dp_multiplier_path` for
    SUM aggregate; per-spin expectation here is a fast approximation).
    """
    e_m = params.initial_multiplier + params.bump_increment * params.p_bump_per_spin * (t - 1)
    if params.max_multiplier is not None:
        e_m = min(e_m, params.max_multiplier)
    return e_m


def _dp_multiplier_path(params: PersistentMultiplierParams) -> list[float]:
    """Exact DP: E[multiplier_t] for t=1..T accounting for cap.

    State: distribution over multiplier values at spin t. Bump operation
    advances multiplier deterministically by `bump_increment` if bump
    fires, else stays. Cap clamps any value at max_multiplier.

    For tractability, we discretize: multiplier values fall on grid
    `initial + k × bump_increment` for k in [0, K_max] where K_max =
    max_bumps_before_cap.
    """
    initial = params.initial_multiplier
    bump = params.bump_increment
    cap = params.max_multiplier
    p = params.p_bump_per_spin
    T = params.fs_initial_spins

    # Determine grid size K (number of bumps before hitting cap)
    if cap is not None and bump > 0:
        K = int((cap - initial) // bump)
        if initial + K * bump > cap + 1e-9:
            K -= 1
    else:
        K = T  # at most T bumps over T spins

    # State vector: probs[k] = P(k bumps so far)
    # Initial: 0 bumps before spin 1
    probs = [0.0] * (K + 1)
    probs[0] = 1.0

    e_per_spin: list[float] = []
    for _ in range(T):
        # E[multiplier this spin] = sum_k probs[k] × m_k (capped)
        e_m = 0.0
        for k, pk in enumerate(probs):
            m = initial + k * bump
            if cap is not None:
                m = min(m, cap)
            e_m += pk * m
        e_per_spin.append(e_m)

        # Advance: each k → with prob (1-p) stay, with prob p go k+1
        new_probs = [0.0] * (K + 1)
        for k, pk in enumerate(probs):
            if k == K:
                # at cap, can't go higher → stays
                new_probs[k] += pk
            else:
                new_probs[k] += pk * (1 - p)
                new_probs[k + 1] += pk * p
        probs = new_probs

    return e_per_spin


def expected_fs_total(params: PersistentMultiplierParams) -> float:
    """E[total FS award × bet] across all FS spins (cap-aware DP)."""
    e_per_spin = _dp_multiplier_path(params)
    sum_e_mult = sum(e_per_spin)
    return params.base_pay_per_spin_x_bet * sum_e_mult


def persistent_multiplier_rtp(params: PersistentMultiplierParams) -> dict:
    """Per-base-spin RTP + audit breakdown."""
    e_per_spin = _dp_multiplier_path(params)
    e_total = params.base_pay_per_spin_x_bet * sum(e_per_spin)
    rtp = params.fs_trigger_p * e_total
    return {
        "rtp_contribution": rtp,
        "fs_trigger_p": params.fs_trigger_p,
        "fs_initial_spins": params.fs_initial_spins,
        "base_pay_per_spin_x_bet": params.base_pay_per_spin_x_bet,
        "initial_multiplier": params.initial_multiplier,
        "bump_increment": params.bump_increment,
        "p_bump_per_spin": params.p_bump_per_spin,
        "max_multiplier": params.max_multiplier,
        "expected_multiplier_per_spin": list(e_per_spin),
        "average_multiplier": sum(e_per_spin) / len(e_per_spin),
        "expected_fs_total_x_bet": e_total,
    }

File: rust-sim/src/kernels/persistent_multiplier.rs
//! W244.19 — persistent_multiplier Rust port.
//!
//! Mirror of `tools/math_dsl/persistent_multiplier.py`. Closed-form DP over
//! (k_bumps, spin_idx) state space with cap clamping. Each spin contributes
//! E[multiplier × base_pay]; total FS award = Σ E[mult_t × base_pay].
//!
//! Used by Sticky Bandits, Mighty Wild, NetEnt Multiplier FS, JTG persistent
//! multiplier features.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PersistentMultiplierParams {
    pub fs_trigger_p: f64,
    pub fs_initial_spins: u32,
    pub base_pay_per_spin_x_bet: f64,
    #[serde(default = "default_initial_multiplier")]
    pub initial_multiplier: f64,
    #[serde(default = "default_bump_increment")]
    pub bump_increment: f64,
    #[serde(default = "default_p_bump_per_spin")]
    pub p_bump_per_spin: f64,
    #[serde(default)]
    pub max_multiplier: Option<f64>,
}

fn default_initial_multiplier() -> f64 { 1.0 }
fn default_bump_increment() -> f64 { 1.0 }
fn default_p_bump_per_spin() -> f64 { 0.30 }

impl PersistentMultiplierParams {
    pub fn validate(&self) -> Result<(), String> {
        if !(0.0..=1.0).contains(&self.fs_trigger_p) {
            return Err(format!(
                "fs_trigger_p {} outside [0,1]",
                self.fs_trigger_p
            ));
        }
        if self.fs_initial_spins < 1 {
            return Err("fs_initial_spins must be ≥ 1".into());
        }
        if self.base_pay_per_spin_x_bet < 0.0 {
            return Err("base_pay_per_spin_x_bet must be ≥ 0".into());
        }
        if self.initial_multiplier < 0.0 {
            return Err("initial_multiplier must be ≥ 0".into());
        }
        if self.bump_increment < 0.0 {
            return Err("bump_increment must be ≥ 0".into());
        }
        if !(0.0..=1.0).contains(&self.p_bump_per_spin) {
            return Err(format!(
                "p_bump_per_spin {} outside [0,1]",
                self.p_bump_per_spin
            ));
        }
        if let Some(cap) = self.max_multiplier {
            if cap < self.initial_multiplier {
                return Err(format!(
                    "max_multiplier {} must exceed initial {}",
                    cap, self.initial_multiplier
                ));
            }
        }
        Ok(())
    }
}

/// E[multiplier on FS spin t (1-indexed)] — fast approximation (no cap DP).
pub fn expected_multiplier_at_spin(params: &PersistentMultiplierParams, t: u32) -> f64 {
    let e_m = params.initial_multiplier
        + params.bump_increment * params.p_bump_per_spin * ((t - 1) as f64);
    if let Some(cap) = params.max_multiplier {
        e_m.min(cap)
    } else {
        e_m
    }
}

/// Exact DP: E[multiplier_t] for t=1..T accounting for cap.
fn dp_multiplier_path(params: &PersistentMultiplierParams) -> Vec<f64> {
    let initial = params.initial_multiplier;
    let bump = params.bump_increment;
    let cap = params.max_multiplier;
    let p = params.p_bump_per_spin;
    let t_max = params.fs_initial_spins;

    // Determine grid size K (number of bumps before hitting cap)
    let k_max: usize = if let Some(c) = cap {
        if bump > 0.0 {
            let mut k = ((c - initial) / bump).floor() as i64;
            if initial + (k as f64) * bump > c + 1e-9 {
                k -= 1;
            }
            std::cmp::max(0, k) as usize
        } else {
            t_max as usize
        }
    } else {
        t_max as usize
    };

    // State vector: probs[k] = P(k bumps so far). Initial: 0 bumps before spin 1.
    let mut probs = vec![0.0_f64; k_max + 1];
    probs[0] = 1.0;

    let mut e_per_spin = Vec::with_capacity(t_max as usize);
    for _ in 0..t_max {
        // E[multiplier this spin] = Σ_k probs[k] × m_k (capped)
        let mut e_m = 0.0_f64;
        for (k, pk) in probs.iter().enumerate() {
            let mut m = initial + (k as f64) * bump;
            if let Some(c) = cap {
                m = m.min(c);
            }
            e_m += pk * m;
        }
        e_per_spin.push(e_m);

        // Advance: each k → with prob (1-p) stay, with prob p go k+1
        let mut new_probs = vec![0.0_f64; k_max + 1];
        for k in 0..=k_max {
            let pk = probs[k];
            if k == k_max {
                new_probs[k] += pk; // at cap → stays
            } else {
                new_probs[k] += pk * (1.0 - p);
                new_probs[k + 1] += pk * p;
            }
        }
        probs = new_probs;
    }

    e_per_spin
}

pub fn expected_fs_total(params: &PersistentMultiplierParams) -> f64 {
    let e_per_spin = dp_multiplier_path(params);
    let sum_e_mult: f64 = e_per_spin.iter().sum();
    params.base_pay_per_spin_x_bet * sum_e_mult
}

#[derive(Debug, Serialize)]
pub struct PersistentMultiplierResult {
    pub rtp_contribution: f64,
    pub fs_trigger_p: f64,
    pub fs_initial_spins: u32,
    pub base_pay_per_spin_x_bet: f64,
    pub initial_multiplier: f64,
    pub bump_increment: f64,
    pub p_bump_per_spin: f64,
    pub max_multiplier: Option<f64>,
    pub expected_multiplier_per_spin: Vec<f64>,
    pub average_multiplier: f64,
    pub expected_fs_total_x_bet: f64,
}

pub fn persistent_multiplier_rtp(params: &PersistentMultiplierParams) -> PersistentMultiplierResult {
    let e_per_spin = dp_multiplier_path(params);
    let sum_e_mult: f64 = e_per_spin.iter().sum();
    let e_total = params.base_pay_per_spin_x_bet * sum_e_mult;
    let rtp = params.fs_trigger_p * e_total;
    let n = e_per_spin.len() as f64;
    PersistentMultiplierResult {
        rtp_contribution: rtp,
        fs_trigger_p: params.fs_trigger_p,
        fs_initial_spins: params.fs_initial_spins,
        base_pay_per_spin_x_bet: params.base_pay_per_spin_x_bet,
        initial_multiplier: params.initial_multiplier,
        bump_increment: params.bump_increment,
        p_bump_per_spin: params.p_bump_per_spin,
        max_multiplier: params.max_multiplier,
        average_multiplier: sum_e_mult / n,
        expected_multiplier_per_spin: e_per_spin,
        expected_fs_total_x_bet: e_total,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sticky_bandits_like() -> PersistentMultiplierParams {
        PersistentMultiplierParams {
            fs_trigger_p: 0.005,
            fs_initial_spins: 10,
            base_pay_per_spin_x_bet: 1.0,
            initial_multiplier: 1.0,
            bump_increment: 1.0,
            p_bump_per_spin: 0.30,
            max_multiplier: None,
        }
    }

    #[test]
    fn expected_multiplier_at_first_spin_equals_initial() {
        let p = sticky_bandits_like();
        let e1 = expected_multiplier_at_spin(&p, 1);
        assert!((e1 - 1.0).abs() < 1e-12);
    }

    #[test]
    fn dp_average_matches_no_cap_linear() {
        // No cap, p=0.3, T=10 → avg multiplier ≈ 1 + 0.3 × (0+1+...+9)/10 = 1 + 0.3 × 4.5 = 2.35
        let p = sticky_bandits_like();
        let r = persistent_multiplier_rtp(&p);
        assert!((r.average_multiplier - 2.35).abs() < 1e-10);
    }

    #[test]
    fn dp_with_cap_clamps_at_max() {
        let p = PersistentMultiplierParams {
            fs_trigger_p: 0.01,
            fs_initial_spins: 100, // many spins, force cap saturation
            base_pay_per_spin_x_bet: 1.0,
            initial_multiplier: 1.0,
            bump_increment: 1.0,
            p_bump_per_spin: 0.50,
            max_multiplier: Some(5.0),
        };
        let r = persistent_multiplier_rtp(&p);
        // After enough spins, average should approach 5 (the cap)
        let last_e_m = *r.expected_multiplier_per_spin.last().unwrap();
        assert!(last_e_m > 4.5);
        assert!(last_e_m <= 5.0 + 1e-12);
    }

    #[test]
    fn rtp_includes_fs_trigger_probability() {
        let p = sticky_bandits_like();
        let r = persistent_multiplier_rtp(&p);
        // 0.005 × 10 × 1 × 2.35 = 0.1175
        assert!((r.rtp_contribution - 0.1175).abs() < 1e-10);
    }

    #[test]
    fn validate_rejects_cap_below_initial() {
        let p = PersistentMultiplierParams {
            fs_trigger_p: 0.01,
            fs_initial_spins: 10,
            base_pay_per_spin_x_bet: 1.0,
            initial_multiplier: 5.0,
            bump_increment: 1.0,
            p_bump_per_spin: 0.3,
            max_multiplier: Some(3.0),
        };
        assert!(p.validate().is_err());
    }
}
